Rewriting the SYCL encoding

OmerSakar opened the pull request:
Checklist: <!-- Please first submit your pull request, you can check the checkboxes later. Feel free to ask for help if you don't know how/where to make changes. -->

- [ ] The wiki is updated in accordance with the changes in this PR. For example: syntax changes, semantics changes, VerCors flags changes, etc.

# PR description

<!-- Describe the motivation of the PR and the changes it introduces. Why is it needed, and what does it change? -->

sakehl committed:
Fix for no sequence indexing

OmerSakar committed:
Merge remote-tracking branch 'fmt/dev' into syql-rq4-patch

OmerSakar committed:
Patch for the SYCL support. Not in a state ready for dev yet.

OmerSakar committed:
Merge remote-tracking branch 'origin/dev' into sycl-patch

OmerSakar committed:
Replace the sycl::mul method to sycl::h::mul.

OmerSakar committed:
Intermediate Commit. This commit partially fixes the context_everywhere propagation to the SYCL kernels. However, but due to issue 924, the context_everywhere propagation does not seem to work (yet).

OmerSakar committed:
ContextEverywhere now also in the contract of the kernel runner, however, it seems that the propagation of contexteverywhere in runner methods seems to be buggy. Part of an issue on github.

OmerSakar committed:
Merge branch 'dev' into sycl-patch

OmerSakar committed:
Rewrote the SYCL support to fix scoping issues. Not entirely finished yet.

This is part of a series of rewrites. This is specifically the rewrite for the basic kernel, have not tested the other tests yet.

OmerSakar committed:
Resolved most SYCL related issues with the new encoding. A few tests are still failing.

OmerSakar committed:
Resolved most SYCL related issues with the new encoding. A few tests are still failing. (forgot to push some changes)

OmerSakar committed:
Added typeargs to CPPClassMethodOrFieldAccess.

OmerSakar committed:
Merge branch 'refs/heads/dev' into sycl-patch

# Conflicts:
#	src/rewrite/vct/rewrite/lang/LangCPPToCol.scala

OmerSakar committed:
formatted files

OmerSakar committed:
Added opaque to CPP.

OmerSakar committed:
new example

OmerSakar committed:
Merge branch 'dev' of github.com:utwente-fmt/vercors into sycl-patch

OmerSakar committed:
committing to work on other laptop

OmerSakar committed:
Pushing examples

OmerSakar committed:
Pushing examples

superaxander commented on bin/silicon:
```diff
@@ -3,10 +3,10 @@ set -e
 (cd "$ROOT"; ./mill -j 0 vercors.main.runScript)
 
-if [[ "$OSTYPE" == "darwin" ]]; then
+#if [[ "$OSTYPE" == "darwin" ]]; then
```
You probably didn't mean to commit this :)

OmerSakar committed:
Basic support build for lane id and warp id for SYCL sub_groups. And Added CPPAttribute to reason about warpsize.

OmerSakar committed:
Merge branch 'refs/heads/dev' into sycl-patch

OmerSakar committed:
Added warpsize and added an invariant for the subgroup functions.

Still todo is what happens when warpsize is not explicitly defined.

OmerSakar committed:
Had changed the prepend and append operators on sequences on accident. Now changed back.

OmerSakar committed:
Weekend Commit: I am continuing with the warp support if the warpsize is not defined explicitly. I believe the option to have it define implicitly to be a power of two is fine. Specifically because we also use the CUDA mapping, where other targets do not have a mapping defined, as per the SYCL standard.

But this commit is not done yet, its just a weekend commit.

OmerSakar committed:
The warpsize are now 16, 32 or 64 big if not specified explicitly.

OmerSakar committed:
Moved some global val/vars to the top and removed a commented method.

OmerSakar committed:
Added the check back that checks whether the accessor permission is less then the permissions specified in the contract.

OmerSakar committed:
Fixed a problem with double accessors and removed a test to check for unsupported local accessors constructors with stack variables.

OmerSakar committed:
A proper fix for two problems:

The double accessor problem in WriteOnReadAccessorWithDoubleAccessors.cpp has been solved by generating permissions.

The failing test in WriteToReadAccessor.cpp where the sycl code (perm in accessor) and the perm in the kernel contract did not match (read in accessor, write in contract).

OmerSakar committed:
Merge branch 'dev' into sycl-patch

# Conflicts:
#	bin/silicon

OmerSakar committed:
Removed a testcase since it is not required anymore. The missing range requirements should not be necessary since the ranges are now properly propagated. Added ShiftLeft.cpp to the test suite

OmerSakar committed:
ShiftGroupLeft.cpp Updated and a necessary flag added.

OmerSakar committed:
Optimized the annotations needed for the AccessorDeclarations.cpp example due to generation of the read permissions. Also continued with the final checks for the ShiftGroupLeft.cpp, not finished yet.

OmerSakar committed:
Workign on detecting whether the id functions have been used in the sg invariant. Currently seems to work, but need to polish it and extract it into a method so that it can be reused later.

OmerSakar committed:
Detecting whether the id functions have been used in the sg invariant. It is still a best effort way of doing this.

OmerSakar committed:
Intermediate commit to save some progress. I found a mistake in my shift_group_left implementation. I am trying to resolve it

superaxander commented on res/universal/res/c/emmintrin.h:
Did you change anything in this file? It seems like it was auto-formatted but it does not look great

OmerSakar replied:
I have reverted it back to match dev.

superaxander commented on res/universal/res/c/opencl.h:
Same for this file

OmerSakar replied:
I have reverted it back to match dev.

superaxander commented on res/universal/res/cpp/sycl/sycl.hpp:
```diff
@@ -1,5 +1,28 @@
 namespace sycl {
 
+  namespace h {
```
Will users ever need to interact with this namespace, if so why is it called `h`?

OmerSakar replied:
Where should this be documented?

superaxander replied:
Well wherever the documentation for using SYCL in VerCors is :P

superaxander commented on res/universal/res/cpp/sycl/sycl.hpp:
```diff
@@ -1,5 +1,28 @@
+  namespace h {
+    /*@
+      ensures (a>0 && b>0) ==> \result > 0;
+      ensures \result == a*b;
```
This one should be unnecessary, unless you want to make this function opaque. (Same goes for the `add` function)

superaxander commented on src/col/vct/col/typerules/CoercingRewriter.scala:
```diff
@@ -1269,13 +1270,16 @@ abstract class CoercingRewriter[Pre <: Generation]()
       case CPPLambdaRef() => e
-      case inv @ CPPInvocation(applicable, args, givenArgs, yields) =>
-        CPPInvocation(applicable, args, givenArgs, yields)(inv.blame)
+      case inv @ CPPInvocation(applicable, args, givenArgs, yields, sginv) =>
```
You might want to coerce the sub-group invariant to resource in case someone enters a non-resource expression in there, or is that checked somewhere else already?

superaxander commented on src/llvm/lib/Util/PallasWrapperUtils.cpp:
When you merge this with `dev` take care not to change anything when resolving conflicts

superaxander commented on src/main/vct/main/stages/Resolution.scala:
```diff
@@ -3,19 +3,10 @@ package vct.main.stages
+import vct.col.ast.{ApplicableContract, Expr, Function, GlobalDeclaration, LLVMGlobalSpecification, Program, VCLLVMFunctionContract, Verification, VerificationContext}
 import vct.col.check.CheckError
 import vct.col.origin.{FileSpanningOrigin, Origin, ReadableOrigin}
+import vct.col.print.Ctx
```
Unused import?

superaxander commented on src/main/vct/main/stages/Resolution.scala:
```diff
@@ -32,14 +23,9 @@ import vct.parsers.ParseResult
+import java.io.{InputStreamReader, OutputStreamWriter, StringReader, StringWriter}
 import java.nio.charset.StandardCharsets
-import java.nio.file.Path
+import java.nio.file.{Path, Paths}
```
Unused import?

superaxander commented on src/main/vct/main/stages/Resolution.scala:
```diff
@@ -32,14 +23,9 @@ import vct.parsers.ParseResult
-  StringReader,
-  StringWriter,
-}
+import java.io.{InputStreamReader, OutputStreamWriter, StringReader, StringWriter}
```
Unused import?

superaxander commented on src/rewrite/vct/rewrite/lang/LangCPPToCol.scala:
```diff
@@ -2,31 +2,27 @@ package vct.rewrite.lang
 import vct.col.resolve.lang.CPP
 import vct.col.rewrite.ParBlockEncoder.ParBlockNotInjective
-import vct.col.rewrite.{Generation, ParBlockEncoder, Rewritten}
+import vct.col.rewrite.{Generation, ParBlockEncoder, Rewriter, Rewritten, SimplifyNestedQuantifiers}
```
Unused imports?

superaxander commented on src/rewrite/vct/rewrite/lang/LangCPPToCol.scala:
```diff
@@ -2,31 +2,27 @@ package vct.rewrite.lang
+import vct.col.rewrite.{Generation, ParBlockEncoder, Rewriter, Rewritten, SimplifyNestedQuantifiers}
 import vct.col.util.AstBuildHelpers.{assignLocal, _}
-import vct.col.util.{AstBuildHelpers, SuccessionMap}
+import vct.col.util.{AstBuildHelpers, Substitute, SuccessionMap}
```
Unused import?

superaxander commented on src/rewrite/vct/rewrite/lang/LangCPPToCol.scala:
```diff
@@ -2,31 +2,27 @@ package vct.rewrite.lang
 import vct.result.VerificationError.{Unreachable, UserError}
 import vct.rewrite.lang.LangSpecificToCol.NotAValue
 
+import scala.Seq
```
Unused import?

superaxander commented on src/rewrite/vct/rewrite/lang/LangCPPToCol.scala:
```diff
@@ -2,31 +2,27 @@ package vct.rewrite.lang
 
+import scala.Seq
 import scala.collection.mutable
+import scala.runtime.BoxedUnit
```
Unused import

superaxander commented on src/rewrite/vct/rewrite/lang/LangCPPToCol.scala:
```diff
@@ -210,6 +214,14 @@ case object LangCPPToCol {
       inv.blame.blame(SYCLItemMethodPreconditionFailed(error.node))
   }
 
+  private case class SYCLMulMethodInvocationBlame()
```
Why introduce this blame when you can simply pass in the PanicBlame directly?

superaxander replied:
If you need to use this in multiple places you could just define it as `def SYCLMulMethodInvocationBlame = PanicBlame(...)`

superaxander commented on src/rewrite/vct/rewrite/lang/LangCPPToCol.scala:
```diff
@@ -706,72 +790,68 @@ case class LangCPPToCol[Pre <: Generation](rw: LangSpecificToCol[Pre])
 
-  val currentlyRunningKernels
-      : mutable.Map[Local[Post], Seq[SYCLAccessor[Post]]] = mutable.Map.empty
+  var syclBufferRangeRelations: Seq[Eq[Post]] = Nil
```
Haven't looked further down yet, but this might make more sense as a ScopedStack? Or can there only ever be a single method with SYCL kernels? It seems weird to me that this is a global thing

superaxander commented on src/rewrite/vct/rewrite/lang/LangCPPToCol.scala:
```diff
@@ -706,72 +790,68 @@ case class LangCPPToCol[Pre <: Generation](rw: LangSpecificToCol[Pre])
-      : mutable.Map[Local[Post], Seq[SYCLAccessor[Post]]] = mutable.Map.empty
+  var syclBufferRangeRelations: Seq[Eq[Post]] = Nil
+
+  val currentlyRunningKernels: mutable.Map[Local[
```
Similarly for this one it sounds to me like this doesn't make sense outside some scope

superaxander commented on src/rewrite/vct/rewrite/lang/LangCPPToCol.scala:
```diff
@@ -1409,20 +1534,136 @@ case class LangCPPToCol[Pre <: Generation](rw: LangSpecificToCol[Pre])
+
+        val defineWarpSize = maybeWarpSize match {
+          case Some(ws) => (
+            Assign(Local[Post](syclWarpSize.get.ref)(ws.o), ws.args.head)(PanicBlame("The assignment of the warpsize should never fail"))(ws.o.where(name="warpSize")),
```
We have a predefined PanicBlame for assigning locals `AssignLocalOk`, there's also the helper function `assignLocal(local, value)`

superaxander commented on src/rewrite/vct/rewrite/lang/LangCPPToCol.scala:
```diff
@@ -1409,20 +1534,136 @@ case class LangCPPToCol[Pre <: Generation](rw: LangSpecificToCol[Pre])
+            implicit val o: Origin = invocation.o
+            val wrpsizVar = new Variable[Post](TCInt())(o.where(name="warpSize"))
+            val wrpsizeProcedure = withResult((result: Result[Post]) => {
+                new Procedure[Post](
```
Since your using default arguments here a lot I recommend using the `procedure` helper function from ASTBuildHelpers

superaxander commented on src/rewrite/vct/rewrite/lang/LangCPPToCol.scala:
```diff
@@ -1409,20 +1534,136 @@ case class LangCPPToCol[Pre <: Generation](rw: LangSpecificToCol[Pre])
+              })
+
+              rw.globalDeclarations.declare(wrpsizeProcedure)
+              val callToWrpsizeProcedure = new ProcedureInvocation[Post](
```
Similarly you can use the `procedureInvocation` helper function here

superaxander commented on src/rewrite/vct/rewrite/lang/LangCPPToCol.scala:
```diff
@@ -1409,20 +1534,136 @@ case class LangCPPToCol[Pre <: Generation](rw: LangSpecificToCol[Pre])
+      implicit val o = accRef.decls.o
+      val indexVar = new Variable[Post](TCInt())(o.where(name = "i"))
+      val blame = PanicBlame("Should never fail. Please report if it does fail")
+      Starall(
```
Here I can recommend using the `starall` helper function which makes the variable for you

superaxander commented on src/rewrite/vct/rewrite/lang/LangCPPToCol.scala:
```diff
@@ -1409,20 +1534,136 @@ case class LangCPPToCol[Pre <: Generation](rw: LangSpecificToCol[Pre])
+            ),
+            body = (Local[Post](indexVar.ref) >= c_const(0)  &&
+                Local[Post](indexVar.ref) < Length(Local[Post](buff.ref))(PanicBlame("Should never be null")))
+            ==> Perm(
```
The canonical way to write read permission is with the `Value` node

superaxander commented on src/rewrite/vct/rewrite/lang/LangCPPToCol.scala:
```diff
@@ -1409,20 +1534,136 @@ case class LangCPPToCol[Pre <: Generation](rw: LangSpecificToCol[Pre])
+            triggers = Seq(
+              Seq(
+                ArraySubscript(
+                  Local[Post](buff.ref),
```
This pattern seems to be everywhere here, you can do `v.get` instead this is an extension function from ASTBuildHelpers that expands to `Local(new DirectRef(v))`

superaxander commented on src/rewrite/vct/rewrite/lang/LangCPPToCol.scala:
```diff
@@ -1548,15 +1802,15 @@ case class LangCPPToCol[Pre <: Generation](rw: LangSpecificToCol[Pre])
     currentDimensionIterVars.clear()
     syclAccessorSuccessor.clear()
     currentThis = None
-
+    syclBufferRangeRelations = Nil
```
These are clear candidates for using a ScopedStack

superaxander commented on src/rewrite/vct/rewrite/lang/LangCPPToCol.scala:
```diff
@@ -1592,40 +1839,44 @@ case class LangCPPToCol[Pre <: Generation](rw: LangSpecificToCol[Pre])
     })
     implicit val o: Origin = kernelDeclaration.o
 
+    gatherBlockStatements = true
+    visitedKernelStatements = mutable.Seq.empty
+    val body = rw.dispatch(kernelDeclaration.body)
+    visitedKernelStatements = mutable.Seq.empty
+    gatherBlockStatements = false
```
Another good candidate for using ScopedStack

superaxander commented on src/rewrite/vct/rewrite/lang/LangCPPToCol.scala:
```diff
@@ -1860,26 +2105,24 @@ case class LangCPPToCol[Pre <: Generation](rw: LangSpecificToCol[Pre])
+                    val localBuff = buffer.generatedVar.get(accO)
+
+                    val rangeIndexLocal = buffer.range.dimensions.map { d =>
+                      d.asInstanceOf[Local[Post]]
```
I guess that the condition that buffer.range.dimensions contains only Locals is preserved elsewhere, I am wondering if it'd be possible to just have the field in SYCLAccessor contain Expr's do you need it to be a local? In any case you could add an assertion to the body of SYCLRange to check that it only contains Local's when it is constructed (then you'd get an error where the mistake is made instead of here)

superaxander commented on src/rewrite/vct/rewrite/lang/LangCPPToCol.scala:
```diff
@@ -1926,159 +2169,41 @@ case class LangCPPToCol[Pre <: Generation](rw: LangSpecificToCol[Pre])
-          )(acc.instanceField.o),
-        )
-      )(acc.instanceField.o),
+      foldStar(Seq(
```
foldStar with only one expression in the sequence

superaxander commented on src/rewrite/vct/rewrite/lang/LangCPPToCol.scala:
```diff
@@ -2422,13 +2623,14 @@ case class LangCPPToCol[Pre <: Generation](rw: LangSpecificToCol[Pre])
 
     // Wait for SYCL kernels that access the buffer to finish executing
     val kernelsToTerminate = currentlyRunningKernels.filter(tuple =>
```
Could make this a `.filter { case (key, (pred, accessors)) =>` to prevent the _.2._1 stuff down below

superaxander commented on src/rewrite/vct/rewrite/lang/LangCPPToCol.scala:
```diff
@@ -2609,41 +2796,98 @@ case class LangCPPToCol[Pre <: Generation](rw: LangSpecificToCol[Pre])
-        Starall(bindings, triggers, removeKernelClassInstancePermissions(body))(
-          s.blame
-        )
+        Starall(
```
could be `s.rewrite(body=removeKernelClassInstancePermissions(body, accs))`

superaxander commented on src/rewrite/vct/rewrite/lang/LangCPPToCol.scala:
```diff
@@ -2609,41 +2796,98 @@ case class LangCPPToCol[Pre <: Generation](rw: LangSpecificToCol[Pre])
+          removeKernelAccessorPermissions(right, accs),
+        )
+
+      case ActionPerm(Local(obj), _)
```
Could be `case ActionPerm(Local(Ref(v)), _)` replacing `obj.decl` with `v`

superaxander commented on src/rewrite/vct/rewrite/lang/LangCPPToCol.scala:
```diff
@@ -2609,41 +2796,98 @@ case class LangCPPToCol[Pre <: Generation](rw: LangSpecificToCol[Pre])
+    implicit val o = e.o
+    e match {
+      case s @ Starall(bindings, triggers, body) =>
+        Starall(bindings, triggers, removeRangeAccesses(body))(s.blame)
```
again could be `s.rewrite(body=removeRangeAccesses(body))`

superaxander commented on src/rewrite/vct/rewrite/lang/LangCPPToCol.scala:
```diff
@@ -2794,4 +3030,215 @@ case class LangCPPToCol[Pre <: Generation](rw: LangSpecificToCol[Pre])
+    }
+    val warpsize = syclWarpSize.getOrElse(throw new SYCLWarpSizeNotDefinedError(inv))
+
+    val predToExhale = Implies(
```
could be `val predToExhale = (c_const(0) < laneid - d) ==> exhalePred`

superaxander commented on src/rewrite/vct/rewrite/lang/LangCPPToCol.scala:
```diff
@@ -2794,4 +3030,215 @@ case class LangCPPToCol[Pre <: Generation](rw: LangSpecificToCol[Pre])
+    val shiftGroupLeftBlame = PanicBlame("The call to shift_group_left should never fail")
+
+    val procedure = withResult((result: Result[Post]) => {
+      new Procedure[Post](
```
use `procedure` helper function

superaxander commented on src/rewrite/vct/rewrite/lang/LangCPPToCol.scala:
```diff
@@ -2794,4 +3030,215 @@ case class LangCPPToCol[Pre <: Generation](rw: LangSpecificToCol[Pre])
+        Then(
+          PreAssignExpression(
+            Local[Post](sglResult.ref),
+            ProcedureInvocation[Post](
```
Use `procedureInvocation` helper function

superaxander commented on src/rewrite/vct/rewrite/lang/LangCPPToCol.scala:
```diff
@@ -2794,4 +3030,215 @@ case class LangCPPToCol[Pre <: Generation](rw: LangSpecificToCol[Pre])
+      case (name: CPPName[Pre], vals) => (name.name, vals)
+      case (local: CPPLocal[Pre], vals) => (local.name, vals)
+      case (a, b) => (a, b)
+    }.groupBy(_._1).map(t => (t._1, t._2.flatMap(_._2).toSet))
```
Why add the unnamed things into this map here? It seems like you're just using this to find names anyway?

superaxander commented on src/rewrite/vct/rewrite/lang/LangSpecificToCol.scala:
```diff
@@ -417,7 +425,12 @@ case class LangSpecificToCol[Pre <: Generation](
+      case global: GlobalThreadId[Pre] => {
+        c.cudaGlobalThreadId(global)}
+      case sgfv: SubGroupFuncValue[Pre] if !cpp.syclSubgroupInvSuccessors.isEmpty && cpp.syclSubgroupInvSuccessors.top.contains(sgfv) =>
+        cpp.syclSubgroupInvSuccessors.top(sgfv)
```
could be `cpp.syclSubgroupInvSuccessors.exists(_.contains(sgfv))` I think

OmerSakar committed:
Fixed shift_group_left. Now working on shift_group_right.

OmerSakar committed:
Revert files back to match the dev branch again. Was accidentally touched.

OmerSakar committed:
Processed some of Alexanders PR comments.

OmerSakar committed:
Warpsize function should not inline. Fixed the ShiftGroupLeft.cpp file and removed some unnecessary annotations from sycl.hpp.

OmerSakar committed:
Removed the ShiftGroupRight.cpp file and replaced it with the ShiftGroupSum that does the three way dance. I added another special variable in the subgroupinvariant names \sgtid.

Also, ShiftGroupSum.cpp works! Just need to prove that the result is correct after the block. I should still look into ShiftGroupLeft.cpp, its subgroup invariant uses one of the id functions but the current code does not detect it for some reason.

OmerSakar committed:
One dim (x dimension) is added up. Now need to redo the proof for dimensions y and z.

OmerSakar committed:
Formal: ShiftGroupSum has been verified, although it currently does have warpsize==32. Lets see whether that can be relaxed. Informally: It works! OMG IT WORKS and its pretty late, but it works!

OmerSakar committed:
Slightly polished the ShiftGroupSum.cpp example. continuing with debugging it

OmerSakar committed:
Adding the reveal keyword to CPP and its invocations.

OmerSakar committed:
Mistakenly swapped two lines in CPPtoCOL, did not compile before.

OmerSakar committed:
optimized the verification a bit

OmerSakar committed:
Added ShiftGroupSum.cpp to SYCLSpec

OmerSakar committed:
Merge branch 'dev' into sycl-patch

# Conflicts:
#	src/llvm/lib/Util/PallasWrapperUtils.cpp
#	src/rewrite/vct/rewrite/lang/LangCPPToCol.scala

OmerSakar committed:
Fixed the AccessorInNDRange.cpp. Ellen had made a mistake in the calculation of the id over threadblocks

OmerSakar committed:
Added warpsizes from 8 to 64 big. Maybe we can try to have warpsizes > 8 but 2^k instead of something concrete, hmmmmmmmmmmm.  That would be the next thing to try. The Warpfunction could yield a k for which it holds that 2^k == warpsize. Still have to check ShiftGroupSum.cpp without setting the warpsize explicitly

OmerSakar committed:
ShiftGrupSum.cpp also rewritten for warpsize between 8 and 64

OmerSakar committed:
Pushing examples to save progress. Working on making the warpsize fully generic

OmerSakar committed:
smartsum in ShiftGroupLeft now verifies, only the accumulate functino left

OmerSakar committed:
ShiftGroupLeft.cpp is now complete. It is very unstable though. Maybe something can be done about it. But to be investigated alter.

OmerSakar committed:
Made warpsize >8 and changed the ShiftGroupLeft.cpp file based on that. Also optimized the verification

OmerSakar committed:
Intermediate commit, the verification of ShiftGroupSum.accumulateResults works, but very slowly. The kernel itself seems to get stuck. Need to debug that.

OmerSakar committed:
Another intermedite commit, fixed the kernel, now accumulate is broken again...

OmerSakar committed:
Verifies, although its unstable

OmerSakar committed:
Prove of ShiftGroupSum.accumulateResults is in ShiftGroupSumHelper.cpp. This is mainly to prevent ANTLR from running out of memory

OmerSakar committed:
Making progress on the MM-sg example

OmerSakar committed:
Making progress on the MM-sg example

OmerSakar committed:
Making progress on the MM-sg example, finished the variant with 2D threadblocks, now to add the broadcast function

OmerSakar committed:
Almost done wiht Matrixmultiplication example with subgroups, one lemma left

OmerSakar committed:
Replaced some literal sequences with constant indexes, e.q., seq{a,b}[0] == a. This helps the solver a bit. These kinds of literal sequences are generated a lot for SYCL and it seems that this small difference makes a difference in the stability of the verification.

OmerSakar committed:
Fixed a mistake in the SimplifyNestedQuantifiers where part of the permissions were removed by one of the rewrite rules.

OmerSakar committed:
Done with the examples

OmerSakar committed:
Working on getting extract to work in SYCL for kernels. The implementation seems finished, but some tests are still failing (which seem unrelated). Also moved the ShiftSum Tests into SYCLFullProgramsSpec.

OmerSakar committed:
Added proof hints for failing verification.

OmerSakar committed:
Fixed last two SYCL examples. A read (wildcard) permission was replaced with 1\2 permissions, which resulted in insufficient permission. I also changed added long/double to CPP as ints. I removed ShiftGroupSum from the SYCLSpec since it takes too long and gets stuck from time to time.

OmerSakar committed:
Added the MMSG example to the slow examples to ignore.

OmerSakar committed:
Added a blame for the framed proof.

OmerSakar committed:
Fixed #1246.

OmerSakar committed:
Move the SG examples to the fullExamples folder. I added simplify rules to inline literalsequences with constant indices. Fixed a problem with kernels not terminating. Fixed a problem where we didn't entirely follow the SYCL standard (see sycl.hpp). Removed some dead code. Added a running token, similar to fork/join, to prevent possible multiple (non-scaled) joins.

OmerSakar committed:
Added integration tests for SYCL subgroups. I am now working on porting the fullExamples using subgroups to proper syntax to solve a previous issue.

OmerSakar committed:
Applied the changed to the syntax to the SG examples. Added VectorAdd and MatrixTranspose but with the extract statement.

OmerSakar committed:
Added the new examples with Extract to the spec

OmerSakar committed:
Forgot to push an example

OmerSakar committed:
Corrected name of file

OmerSakar committed:
Formatted files to count lines.
